### PR TITLE
feat(ui): Use SDFs in LineShader

### PR DESCRIPTION
**Pull request by Koranir**
## Summary
This PR makes LineShader use an SDF in the shader, for clearer lines and rounding.

NOTE: As lines don't have a fade out effect anymore, lines are now thicker than the were before, so some UI elements might need to be changed.

## Screenshots
Before:
image
After:
image
Before:
image
After:
image
Before:
image
After:
image

## Performance Impact
This PR does impact the performance, but shouldn't be by too much.

### CODE REVIEW

File: data/_ui/interfaces.txt
[the remark this answers is not in the document]
Comment (Koranir) [reply data/_ui/interfaces.txt:1]: Whoops, my editor autoformatted it :O

**Comment by tibetiroka** [comment]
Can you add a "before" version of the screenshots?

**Comment by Koranir** [comment]
> You know, it would not be impossible to add fade.

It's more that the fade was causing the problems in the first place (for example, if you look at the wormhole direction arrows in the first screenshot, you can see how much clearer they are). The fade was used as a sort of anti-aliasing, but the effect really breaks down at larger scales.

**Comment by Koranir** [comment]
> Definitely looks good, any performance tests?

If I shove a bunch of lines onto my screen (500+) and run, I get about ~24% GPU usage, compared to around ~22% GPU usage pre-PR. It's small enough and lines are easy enough to draw that there aren't going to be any problems imo.

**Comment by warp-core** [comment]
Is there any particular reason for changing the format of the OpenGL shader code?

**Comment by Koranir** [comment]
> Is there any particular reason for changing the format of the OpenGL shader code?

It's easier to edit/debug this way.

**Comment by warp-core** [comment]
> > Is there any particular reason for changing the format of the OpenGL shader code?
> 
> It's easier to edit/debug this way.

If you want to propose such a change, it should probably be done while not modifying the functionality of the code.

File: source/LineShader.h [hunk 1]
```diff
@@ -23,9 +23,15 @@ class Point;
 // Class to be used for drawing lines. The sides of a line are anti-aliased, but
 // the start and end of the line are not.
 class LineShader {
+public:
+	enum class Cap : int {
+		Solid = 0,
+		Rounded = 1,
+	};
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/LineShader.h:30]: Do we even need the explicit enum values of 0 and 1?

File: source/LineShader.cpp [hunk 2]
```diff
@@ -42,41 +43,74 @@ void LineShader::Init()
...
+		"uniform int cap;\n"
+		"\n"
+		"in vec2 pos;\n"
 		"out vec4 finalColor;\n"
-
+		"\n"
+		"float udSegment(vec2 p, vec2 a, vec2 b) {\n"
+		"    vec2 ba = b-a;\n"
+		"    vec2 pa = p-a;\n"
+		"    float h = clamp(dot(pa,ba)/dot(ba,ba), 0.0, 1.0);\n"
```
[the remark this answers is not in the document]
Comment (Koranir) [reply source/LineShader.cpp:83]: Normalising the vectors needs two divisions and square roots, this way only needs one division.

File: source/LineShader.cpp [hunk 3]
```diff
@@ -42,41 +43,93 @@ void LineShader::Init()
...
+		"    float dist;\n"
+		"    if(cap == 1) {\n"
+		// Rounded caps can shortcut to a segment sdf.
+		// Segment sdf only provides a distance fromt the line itself so we manually subtract it from the width.
+		"        dist = width - sdSegment(pos, start, end);\n"
+		"    } else {\n"
+		// Suubtract from 1 here to add some AA.
+		"        dist = 1. - sdOrientedBox(pos, start, end, width);\n"
+		"    }\n"
+		"    float alpha = clamp(dist, 0.0, 1.0);\n"
```
[the remark this answers is not in the document]
Comment (Koranir) [reply source/LineShader.cpp:122]: No, alpha values above 1 blow out colours on larger lines:
Without clamp:
image
With clamp:
image

File: source/LineShader.cpp [hunk 4]
```diff
@@ -42,41 +43,91 @@ void LineShader::Init()
...
-		"  tscale = length(len);\n"
-		"  gl_Position = vec4((start + vert.x * len + vert.y * width) * scale, 0, 1);\n"
+		// Construct a rectangle around the line that can accommodate a line of width "width".
+		"    vec2 unit = normalize(end - start);\n"
+		// The vertex will originate from the start or endpoint of the line, depending on the input vertex data.
+		"    vec2 origin = vert.y > 0.0 ? start : end;\n"
+		// Pad the width by 1 so the SDFs have enough space to naturally anti-alias.
+		"    float widthOffset = width + 1;\n"
+		// If the cap is rounded, offset along the unit vector by the width, as the cap is circular with radius
+		//     'width' from the start/endpoints. This is also padded by 1 to allow for anti-aliasing.
```
Comment (tibetiroka) [inline source/LineShader.cpp:65]:
```suggestion
		//     "width" from the start/endpoints. This is also padded by 1 to allow for anti-aliasing.
```

**Review by tibetiroka — COMMENTED** [review]
When I click on the volume bar, the line extends a couple (~5) pixels above the cursor.

**Comment by tibetiroka** [comment]
Requesting @Saugia's opinion on two points:
 - Do you like the visual changes?
 - Do you think the dark box around the volume slider should be changed to have rounded ends to match the volume slider?

**Comment by Koranir** [comment]
> When I click on the volume bar, the line extends a couple (~5) pixels above the cursor.

This is an issue in the base game, but I've fixed it in this PR.

**Comment by tibetiroka** [comment]
Still seems to be a couple pixels off, but it's much better now.

**Comment by Koranir** [comment]
> Still seems to be a couple pixels off, but it's much better now.

This might be because the rounding of the cap in the line has some padding to make round. The center of the rounded cap is where the actual value is. This should make it more in line with the intuitive behaviour. The problem, I think, is that dragging the bar doesn't update it smoothly, which really highlights these non-perfect alignments.

**Pull request merged**